src/app/AttributeAccessInterfaceRegistry.h

```diff
@@ -16,39 +16,49 @@
 #pragma once
 
 #include <app/AttributeAccessInterface.h>
-#include <app/util/af-types.h>
+#include <app/AttributeAccessInterfaceCache.h>
 
-/**
- * Register an attribute access override.  It will remain registered until the
- * endpoint it's registered for is disabled (or until shutdown if it's
- * registered for all endpoints) or until it is explicitly unregistered.
- * Registration will fail if there is an already-registered override for the
- * same set of attributes.
- *
- * @return false if there is an existing override that the new one would
- *               conflict with.  In this case the override is not registered.
- * @return true if registration was successful.
- */
-bool registerAttributeAccessOverride(chip::app::AttributeAccessInterface * attrOverride);
+namespace chip {
+namespace app {
 
-/**
- * Unregister an attribute access override (for example if the object
- * implementing AttributeAccessInterface is being destroyed).
- */
-void unregisterAttributeAccessOverride(chip::app::AttributeAccessInterface * attrOverride);
+class AttributeAccessInterfaceRegistry
+{
+public:
+    /**
+     * Register an attribute access override.  It will remain registered until the
+     * endpoint it's registered for is disabled (or until shutdown if it's
+     * registered for all endpoints) or until it is explicitly unregistered.
+     * Registration will fail if there is an already-registered override for the
+     * same set of attributes.
+     *
+     * @return false if there is an existing override that the new one would
+     *               conflict with.  In this case the override is not registered.
+     * @return true if registration was successful.
+     */
+    bool Register(AttributeAccessInterface * attrOverride);
 
-/**
- * Unregister all attribute access interfaces that match this given endpoint.
- */
-void unregisterAllAttributeAccessOverridesForEndpoint(EmberAfDefinedEndpoint * definedEndpoint);
+    /**
+     * Unregister an attribute access override (for example if the object
+     * implementing AttributeAccessInterface is being destroyed).
+     */
+    void Unregister(AttributeAccessInterface * attrOverride);
 
-namespace chip {
-namespace app {
+    /**
+     * Unregister all attribute access interfaces that match this given endpoint.
+     */
+    void UnregisterAllForEndpoint(EndpointId endpointId);
 
-/**
- *  Get the registered attribute access override. nullptr when attribute access override is not found.
- */
-AttributeAccessInterface * GetAttributeAccessOverride(EndpointId aEndpointId, ClusterId aClusterId);
+    /**
+     *  Get the registered attribute access override. nullptr when attribute access override is not found.
+     */
+    AttributeAccessInterface * Get(EndpointId aEndpointId, ClusterId aClusterId);
+
+    static AttributeAccessInterfaceRegistry & Instance();
+
+private:
+    AttributeAccessInterface * mAttributeAccessOverrides = nullptr;
```

Comment:
Is this always intended to be a singleton? If so, would it make sense to hide the constructor in private?

Reply:
Used as a singleton in the app, however I am unsure if we ever need testing. I have been having a hard time working with code that went too far into protecting from the user (usually with many private things and having friend only access) so I generally default to public for testability unless there are very strong reasons to do otherwise.

Here it probably does not matter as much as we could `clear all` in tests, however I don't see too much of a harm if we allow constructors. Overally globals are not ideal. In the future I would rather expect we want CodegenDataModelProvider to own these if possible (not possible today because all cluster implementations are coupled and assume a global).